### PR TITLE
op CALLDATALOAD specs

### PR DESCRIPTION
**Pull request by roynalnaruto**

### CODE REVIEW

File: src/zkevm_specs/evm/execution/calldataload.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+        src_addr = offset
+        src_addr_end = calldata_length
+    else:
+        calldata_length = instruction.call_context_lookup(CallContextFieldTag.CallDataLength)
+        calldata_offset = instruction.call_context_lookup(CallContextFieldTag.CallDataOffset)
+        src_addr = offset + calldata_offset
+        src_addr_end = calldata_offset + calldata_length
+
+    bytes_left = N_BYTES_WORD if calldata_length.n > src_addr_end.n else src_addr_end - src_addr
+    print("bytes left = ", bytes_left)
```
Comment (ChihChengLiang) [inline src/zkevm_specs/evm/execution/calldataload.py:29]:
```suggestion
```

File: src/zkevm_specs/evm/execution/calldataload.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+        calldata_offset = 0
+        src_addr = offset
+        src_addr_end = calldata_length
+    else:
+        calldata_length = instruction.call_context_lookup(CallContextFieldTag.CallDataLength)
+        calldata_offset = instruction.call_context_lookup(CallContextFieldTag.CallDataOffset)
+        src_addr = offset + calldata_offset
+        src_addr_end = calldata_offset + calldata_length
+
+    bytes_left = N_BYTES_WORD if calldata_length.n > src_addr_end.n else src_addr_end - src_addr
```
Comment (ChihChengLiang) [inline src/zkevm_specs/evm/execution/calldataload.py:28]: Can we express this in `instruction.select`?
And can we use `src_addr_end` for consistency with the circuit?

File: src/zkevm_specs/evm/execution/calldataload.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    print("bytes left = ", bytes_left)
+    buffer_reader = BufferReaderGadget(
+        instruction, N_BYTES_WORD, src_addr, src_addr_end, bytes_left
+    )
+
+    calldata_word = []
+    for idx in range(32):
+        if buffer_reader.read_flag(idx):
+            if instruction.curr.is_root:
```
Comment (ChihChengLiang) [inline src/zkevm_specs/evm/execution/calldataload.py:37]: Should we express these in select too?

**Review by ChihChengLiang — COMMENTED** [review]
First round of review

File: src/zkevm_specs/evm/execution/calldataload.py (hunk 2 above)
Comment (roynalnaruto) [reply src/zkevm_specs/evm/execution/calldataload.py:28]: I tried using `instruction.select` for a few cases like:
```python
calldata_length = instruction.select(
    instruction.curr.is_root,
    instruction.tx_context_lookup(tx_id, TxContextFieldTag.CallDataLength),
    instruction.call_context_lookup(CallContextFieldTag.CallDataLength),
)
```

But this tries to add those two lookups before passing the looked up values to the `select` function.

This leads to lookup failures.

Comment (roynalnaruto) [reply src/zkevm_specs/evm/execution/calldataload.py]: I've used the `N_BYTES_WORD` to match the circuits implementation. Thanks for noticing that :)

File: src/zkevm_specs/evm/execution/calldataload.py (hunk 3 above)
Comment (roynalnaruto) [reply src/zkevm_specs/evm/execution/calldataload.py:34]: Similar issue as I mentioned above. We would need an `instruction.condition` method that conditionally adds constrains like here:
```rust
cb.condition(my_conditional_expression, |cb| {
    // my constraints
});
```

File: src/zkevm_specs/evm/execution/calldataload.py (hunk 2 above)
Comment (ChihChengLiang) [reply src/zkevm_specs/evm/execution/calldataload.py]: No problem :)
Your way sounds good.

**Review by ChihChengLiang — APPROVED** [review]
LGTM

File: specs/opcode/35CALLDATALOAD.md [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+
+## Constraints
+
+1. opId == 0x35
+2. State Transition:
+   - gc + 3 (1 stack read, 1 call context read, 1 stack write)
+   - stack_pointer unchanged
+   - pc + 1
+   - gas - 3
+3. Lookups:
```
Comment (CPerezz) [inline specs/opcode/35CALLDATALOAD.md:28]: This is also missing the Lookup of reading the `offset` from the Stack

File: specs/opcode/35CALLDATALOAD.md [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+Stack input is the byte offset to read call data from. Stack output is a 32-byte value starting from the given offset of the call data. All bytes after the end of the call data are set to `0`.
+
+## Circuit Behaviour
+
+1. Do busmapping lookup for stack read operation.
+2. Construct call context table in RW table (for `tx.id`).
+3. Construct tx context table in RW table (for every single byte of `tx.calldata`).
+4. Do busmapping lookup for call context `txid` read operation.
+5. Do busmapping lookup for tx context `calldata` read operation (for each one of the 32 bytes of the stack output).
+6. Do busmapping lookup for stack write operation.
```
Comment (CPerezz) [inline specs/opcode/35CALLDATALOAD.md:18]: Missing Stack Read operation lookup for the `offset`.

File: src/zkevm_specs/evm/execution/calldataload.py [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+    else:
+        calldata_length = instruction.call_context_lookup(CallContextFieldTag.CallDataLength)
+        calldata_offset = instruction.call_context_lookup(CallContextFieldTag.CallDataOffset)
+
+    src_addr = offset + calldata_offset
+    src_addr_end = calldata_length + calldata_offset
+
+    buffer_reader = BufferReaderGadget(
+        instruction, N_BYTES_WORD, src_addr, src_addr_end, N_BYTES_WORD
+    )
```
Comment (CPerezz) [inline src/zkevm_specs/evm/execution/calldataload.py:29]: I'm not sure about this part TBH.

As per Geth impl: https://github.com/ethereum/go-ethereum/blob/master/core/vm/instructions.go#L283-L292

At no point we require to read from the context CALLDATALENGTH or CALLDATAOFFSET.

The paper and the impl just say try to get the 32 bytes starting at offset. If not enough, right pad for each one.

File: src/zkevm_specs/evm/execution/calldataload.py [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+                tx_byte = instruction.tx_calldata_lookup(tx_id, offset + idx)
+                buffer_reader.constrain_byte(idx, tx_byte)
+                calldata_word.append(int(tx_byte))
+            else:
+                mem_byte = instruction.memory_lookup(RW.Read, offset + idx)
+                buffer_reader.constrain_byte(idx, mem_byte)
+                calldata_word.append(int(mem_byte))
+        else:
+            buffer_reader.constrain_byte(idx, 0)
+            calldata_word.append(0)
```
Comment (CPerezz) [inline src/zkevm_specs/evm/execution/calldataload.py:44]: I'm not sure you can append here directly. The padding is done with 0's on the left (If we consider BigEndian). As you se here: https://github.com/ethereum/go-ethereum/blob/8fddf27a989e246659fd018ea9be37b2b4f55326/core/vm/common.go#L54 (This is with LE and bytes).

**Review by CPerezz — CHANGES_REQUESTED** [review]
Some questions related to the fields used in the opcode impl.

Not really sure the logic is correct. But happy to see an explanation as maybe I'm wrong or not getting it :)

File: src/zkevm_specs/evm/execution/calldataload.py (hunk 6 above)
Comment (CPerezz) [reply src/zkevm_specs/evm/execution/calldataload.py:29]: Might this be a result of some copying from CALLDATACOPY?

Comment (ChihChengLiang) [reply src/zkevm_specs/evm/execution/calldataload.py:29]: I agree with @CPerezz. No need to lookup call_context for CallDataLength and CallDataOffset

File: src/zkevm_specs/evm/execution/calldataload.py [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+
+    calldata_word = []
+    for idx in range(N_BYTES_WORD):
+        if buffer_reader.read_flag(idx):
+            if instruction.curr.is_root:
+                tx_byte = instruction.tx_calldata_lookup(tx_id, offset + idx)
+                buffer_reader.constrain_byte(idx, tx_byte)
+                calldata_word.append(int(tx_byte))
+            else:
+                mem_byte = instruction.memory_lookup(RW.Read, offset + idx)
```
Comment (ed255) [inline src/zkevm_specs/evm/execution/calldataload.py:39]: It's not clear to me that this is the right approach.  Shouldn't this be using the `src_addr` instead of `offset` and using caller callID instead of current callID?

File: src/zkevm_specs/evm/execution/calldataload.py (hunk 6 above)
Comment (CPerezz) [reply src/zkevm_specs/evm/execution/calldataload.py:29]: After listening to @ed255 I agree with what he said. I missed the fact that:

>after a call, the call data is copied from the caller memory into a buffer at scope.Contract.Input, but we don't do that
>[author=ed255]

File: src/zkevm_specs/evm/execution/calldataload.py (hunk 7 above)
Comment (roynalnaruto) [reply src/zkevm_specs/evm/execution/calldataload.py:44]: My understanding is that if the buffer reader has marked that index as "no data available", that needs to be padded with `0` (in the case of `CALLDATALOAD`), hence:
```python
if buffer_reader.read_flag(idx):
    // append from tx calldata (root) or memory (internal)
else:
    // append 0 (this is the right padding)
```

File: specs/opcode/35CALLDATALOAD.md (hunk 5 above)
Comment (roynalnaruto) [reply specs/opcode/35CALLDATALOAD.md:18]: I have updated the spec document, I had forgotten to update it after changing the opcode implementation.

File: specs/opcode/35CALLDATALOAD.md (hunk 4 above)
Comment (roynalnaruto) [reply specs/opcode/35CALLDATALOAD.md:28]: I have updated the spec document, I had forgotten to update it after changing the opcode implementation.

File: specs/opcode/35CALLDATALOAD.md [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+## Constraints
+
+1. opId == 0x35
+2. State Transition:
+   - gc + 3 (1 stack read, 1 call context read, 1 stack write)
+   - stack_pointer unchanged
+   - pc + 1
+   - gas - 3
+3. Lookups:
+   - for index `i in range(32)`: `stack_top[i]` is in the RW table {tx context, call data, i}
```
[the remark this answers is not in the document]
Comment (roynalnaruto) [reply specs/opcode/35CALLDATALOAD.md]: There's actually:
- 1 stack read for `offset`
- 1 stack write for `calldata_word`
- i ∈ {0, 1, ..., 31, 32} lookups to the tx table (tag=TxTableFieldTag.CallData) depending on `src_addr` and `src_addr_end`

File: src/zkevm_specs/evm/execution/calldataload.py (hunk 8 above)
Comment (roynalnaruto) [reply src/zkevm_specs/evm/execution/calldataload.py]: Yes, I've updated this to be the `src_addr`. Also added another test case where call data offset is non-zero.

> using caller callID instead of current callID
Is this needed? My understanding was that the current call ID is sufficient. But I may be wrong, could you also help reviewing this part @icemelon ?

File: specs/opcode/35CALLDATALOAD.md [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+     - rw_counter += rw_counter_offset ∈ {4, 5, ..., 35, 36} (1 stack read, 2 call context reads, i ∈ {0, 1, ..., 31, 32} memory reads, 1 stack write)
+   - stack_pointer unchanged
+   - pc + 1
+   - gas - 3
+3. Lookups:
+   - `offset` is at the top of the stack
+   - `tx_id` is in the RW table (call context)
+   - if is_root_call (where `src_addr = offset`):
+     - `calldata_length` is in the TX table
+     - i ∈ {0, 1, ..., 31, 32} lookups for `i in range(32)`: if `buffer.read_flag(i)` then the i'th byte of the element on top of the stack `calldata_word[i]` is in the TX table {tx id, call data, src_addr + i}
```
Comment (DreamWuGit) [inline specs/opcode/35CALLDATALOAD.md:27]: should be i ∈ {0, 1, ..., 31} ?

File: specs/opcode/35CALLDATALOAD.md [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+   - if is_internal_call (where `src_addr = offset + calldata_offset`):
+     - `calldata_length` is in the RW table (call context)
+     - `calldata_offset` is in the RW table (call context)
+     - i ∈ {0, 1, ..., 31, 32} lookups for `i in range(32)`: if `buffer.read_flag(i)` then the i'th byte of the element on top of the stack `calldata_word[i]` is in the RW table {memory, src_addr + i}
```
Comment (DreamWuGit) [inline specs/opcode/35CALLDATALOAD.md:31]: the same as above

File: src/zkevm_specs/evm/execution/calldataload.py [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+    instruction.constrain_equal(
+        instruction.stack_push(),
+        instruction.bytes_to_rlc(bytes(calldata_word)),
+    )
+
+    instruction.step_state_transition_in_same_context(
+        opcode,
+        rw_counter=Transition.delta(instruction.rw_counter_offset),
+        program_counter=Transition.delta(1),
+        stack_pointer=Transition.delta(0),
```
Comment (DreamWuGit) [inline src/zkevm_specs/evm/execution/calldataload.py:55]: better to  use `stack_pointer=Transition.same(),`

File: src/zkevm_specs/evm/execution/calldataload.py (hunk 8 above)
Comment (ed255) [reply src/zkevm_specs/evm/execution/calldataload.py]: I'll try to write an example to show why I think caller_id must be used:

Imagine the following situation:
from call A we do a call B.  A is the calleer, B is the callee.  When doing the call B, A sets some call data [3,4].  It would look like this:

| Step | Curr. Call | operation                  | curr. mem |
| ---- | ---------- | -------------------------- | --------- |
| 0    | A          | `MSTORE(1, 3)`             | `[]`      |
| 1    | A          | `MSTORE(2, 4)`             | `[0,3]`   |
| 2    | A          | `CALL(_, _, _, 2, 2, ...)` | `[0,3,4]` |
| 3    | B          | `CALLDATALOAD(1)`          | `[]`      |

When entering a new call, we are in a new context so we start with empty memory.  Now in step 3, if we read the current call memory we'll find it empty.  It's in the parent's call (B) memory were we'll find the call data.

File: specs/opcode/35CALLDATALOAD.md (hunk 10 above)
Comment (roynalnaruto) [reply specs/opcode/35CALLDATALOAD.md:27]: For the case where no byte is actually read (`out of bound` or `calldatalength=0`, etc.) we have `0` lookups for from the tx table / call context. But if we read 32 bytes, there are `32` lookups. `i` is not the index of the byte being looked up, but the number of lookups, hence it belongs to a set of `{0, 1, 2, ..., 31, 32}` here.

File: src/zkevm_specs/evm/execution/calldataload.py (hunk 8 above)
Comment (roynalnaruto) [reply src/zkevm_specs/evm/execution/calldataload.py]: > It's in the parent's call (B) memory

@ed255 you mean `It's in the parent's call (A) memory` , right?

File: src/zkevm_specs/evm/execution/calldataload.py (hunk 7 above)
Comment (roynalnaruto) [reply src/zkevm_specs/evm/execution/calldataload.py:44]: @CPerezz Hmm, I am a little confused because I had followed the playground [setup here](https://www.evm.codes/playground?unit=Wei&callData=0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF&codeType=Mnemonic&code=%27~1w0yzz~2w31y%27~%2F%2F%20Example%20z%5CnyzCALLDATALOADwzPUSH1%20%01wyz~_).

Could you also have a look at it?

File: src/zkevm_specs/evm/execution/calldataload.py (hunk 8 above)
Comment (icemelon) [reply src/zkevm_specs/evm/execution/calldataload.py]: Yes, I made the same error in the `CALLDATACOPY` that uses the callee's call id for the memory read. Could you also fix it in this PR?

File: specs/opcode/35CALLDATALOAD.md (hunk 10 above)
Comment (CPerezz) [reply specs/opcode/35CALLDATALOAD.md:27]: Ohhh this was num of lookups.

Sorry, I saw `i` used later as the range index and I thought that it was not meaning num of lookups.

File: src/zkevm_specs/evm/execution/calldataload.py (hunk 7 above)
Comment (CPerezz) [reply src/zkevm_specs/evm/execution/calldataload.py:44]: kk it's my bad again. The values are padded correctly as for LE you padd to the right obviously..

Thanks for the link @roynalnaruto and sorry for the noise! :smile:

File: src/zkevm_specs/evm/execution/calldataload.py (hunk 8 above)
Comment (ed255) [reply src/zkevm_specs/evm/execution/calldataload.py]: > > It's in the parent's call (B) memory
> 
> @ed255 you mean `It's in the parent's call (A) memory` , right?

yeah! sorry!  I've edited my original message to fix that.

Comment (roynalnaruto) [reply src/zkevm_specs/evm/execution/calldataload.py]: @ed255 I've made the change in the latest commit. Could you review again?

File: tests/evm/test_calldataload.py [hunk 13]
```diff
@@ -0,0 +1,227 @@
...
+    call_data: bytes,
+    call_data_length: U64,
+    offset: U64,
+    expected_stack_top: bytes,
+    is_root: bool,
+    call_data_offset: Optional[U64],
+):
+    randomness = rand_fp()
+
+    tx = Transaction(id=1, call_data=call_data)
```
Comment (icemelon) [inline tests/evm/test_calldataload.py:75]: only include `call_data` when `is_root` is true.

**Review by icemelon — APPROVED** [review]
otherwise lgtm.

File: tests/evm/test_calldataload.py (hunk 13 above)
Comment (roynalnaruto) [reply tests/evm/test_calldataload.py:75]: Yes I might have missed this after the recent changes. Thanks for pointing out!

**Comment by roynalnaruto** [comment]
@ChihChengLiang @ed255 @CPerezz would be great to have your reviews on the circuits PR https://github.com/appliedzkp/zkevm-circuits/pull/346

File: src/zkevm_specs/evm/execution/calldataload.py (hunk 6 above)
Comment (DreamWuGit) [reply src/zkevm_specs/evm/execution/calldataload.py]: >in geth, after a call, the call data is copied from the caller memory into a buffer at scope.Contract.Input,   

@ed255  appreciate if you can share the code snippet doing this  . I only found `Contract.Input` is assigned here https://github.com/ethereum/go-ethereum/blob/master/core/vm/interpreter.go#L164

Comment (ed255) [reply src/zkevm_specs/evm/execution/calldataload.py]: > > in geth, after a call, the call data is copied from the caller memory into a buffer at scope.Contract.Input,
> 
> @ed255 appreciate if you can share the code snippet doing this . I only found `Contract.Input` is assigned here https://github.com/ethereum/go-ethereum/blob/master/core/vm/interpreter.go#L164

Sure!  Here's an example for the `CALL`:
1. `args` has a `[]byte` slice with the call data here: https://github.com/ethereum/go-ethereum/blob/538a86838423add89a0734d5334599898be91e5a/core/vm/instructions.go#L677 (technically it's not a copy of the call data, as in go, a slice is like a view into a possibly bigger array). This is passed to `interpreter.evm.Call` a few lines below.
2. In `interpreter.evm.Call`, the `args` is received as an argument called `input` and is passed to a function `evm.interpreter.Run`: https://github.com/ethereum/go-ethereum/blob/538a86838423add89a0734d5334599898be91e5a/core/vm/evm.go#L228
3. In `evm.interpreter.Run` we reach the line you mentioned, where `input` is assigned to `contract.Input` which is the line you shared.

So that's how the call data is set in the value `contract.Input`.

**Review by ed255 — APPROVED** [review]
LGTM!  Good work!
